### PR TITLE
Eng 1241 Add ability to fetch an artifact from an existing flow version

### PR DESCRIPTION
**Pull request by Fanjia-Yan**
## Describe your changes and why you are making these changes
This PR add `get(artifact_name)` feature for `flow` object. By passing in an artifact name, the function will return the Artifact associated with the artifact_name. If not, raise ArtifactNotFound Error.

I have created a integration test:
The test create a workflow and publish it online. After the workflow has finished, I use get() to extract the Artifact provided the artifact name. Finally, I checked if the returned Artifact object has the same name as provided.

## Related issue number (if any)
Eng 1241
## Checklist before requesting a review
- [x ] I have performed a self-review of my code.
- [x ] If this is a new feature, I have added unit tests and integration tests.
- [x ] I have manually run the integration tests and they are passing.
- [ ] All features on the UI continue to work correctly.

## Update
Many changes to this PR after last review
1. Complete the `get_workflow()` so that the `workflowresponse` also has serialized function attached to each operator. To get the serialized function, I route "/api/function/%s/export" and request by operator_id
2. Modify the previous `flow.get()` into `flow.artifact()`. In `flow.artifact()`, we first fetch the `artifact.Artifact` object with matching `artifact_name` and convert it to `generic_artifact.Artifact` so that we can call `.get()`.
3. For the integration test. I fetch the artifact using `flow.artifact(output_artifact.name())`. And check for name equality and dataframe equality between locally created artifact and fetched from server artifact.

## Update 2.0
1. Migrate `flow.artifact()` to `flowrun.artifact()`
2. Create a `export_serialized_function()` to handle the getting serialized function into the dag_response rather than putting them all in `get_workflow()`
3. As mentioned by Kenny, the operation on `operator` is crossing the the abstraction barrier. Therefore, in the `api_client`, I eliminated any operation on `operator`  and put them into `workflow_dag_response`
4. In order to prevent reuse of `artifact` from `flowrun.arifact()`, I create a boolean parameter for Artifact(TableArtifact etc.) `from_flow_run`. The parameter defaults to false but can be set to true when creating the Artifact Object. Then, in `decorator.py`, when the operators and artifacts are wrapped, I check whether any input artifact `from_flow_run` is set to true. If so return a Exception.
5. I also create a test case to test whether we catch any input artifact from flow run. I basically publish a flow and fetch the artifact and reuse it for sentiment model. When feeding in the return artifact, we are expected an Exception.

## Update 3.0
After last review, several things have been changed:
1. the process of updating operator will be taking place in `dag.py` instead of `api_client.py`
2. the action of putting serialized function into the dag will be taking place in `flow.py: construct_flow_run()` instead of `api_client.py: get_workflow()`
3. fixing some misc and updating documentations for `_from_flow_run` parameter for InputArtifact

This version looks nicer since serveral lines of codes are taken away from the api_client file :)

### CODE REVIEW

File: integration_tests/sdk/flow_test.py
Comment (vsreekanti) [inline integration_tests/sdk/flow_test.py:176]: Tiny nit here -- shouldn't this be `test_get_artifact_from_flow` or something similar?

File: integration_tests/sdk/flow_test.py
Comment (vsreekanti) [inline integration_tests/sdk/flow_test.py:190]: Is there any other metadata returned that we can check other than the name of the artifact?

File: sdk/aqueduct/flow.py
Comment (vsreekanti) [inline sdk/aqueduct/flow.py:158]: `get()` feels a little ambiguous to me. Are we getting code, data, metadata? Feels like a generic `get()` method should support all of those. I would suggest doing something specific like either `flow.artifact(name)` or `flow.get_artifact(name)`. 

@kenxu95, any thoughts?

File: sdk/aqueduct/flow.py
Comment (vsreekanti) [inline sdk/aqueduct/flow.py:164]: Are we assuming here that when we have a handle to a `flow`, it always points to the latest version @kenxu95?

Comment (vsreekanti) [reply sdk/aqueduct/flow.py:164]: I guess I'm really asking how this interacts with versioning -- can we access older versions?

**Review by vsreekanti — COMMENTED** [review]
@Fanjia-Yan, this mostly looks good, but I left a few comments and nits inline.

Let's also please make sure to title our pull requests clearly -- GitHub defaults to pulling from the branch name, but that's a little unhelpful/unclear in this case.

Thanks!

File: sdk/aqueduct/flow.py
Comment (kenxu95) [reply sdk/aqueduct/flow.py:158]: I think `flow.artifact(name)` is consistent with how we usually do things.

File: sdk/aqueduct/flow.py
Comment (kenxu95) [reply sdk/aqueduct/flow.py:164]: yeah, @Fanjia-Yan  this method shouldn't be on the flow, this should be on the flow run. A flow itself is actually dag-agnostic right now, since a dag is always associated with a particular run. You can also grab a run with `flow.latest()` or `flow.fetch()`

**Review by kenxu95 — CHANGES_REQUESTED** [review]
So checking the artifact name is fine, but it's incomplete. We should also make sure that calling `artifact.get()` returns the same value as before! An artifact fetched from a flow should be functionally indistinguishable from an artifact that was locally created.

I think thats a non-trivial follow up since you'll have to alter the `get_workflow` route to also return the serialized functions.

**Comment by Fanjia-Yan** [comment]
Many changes to this PR after last review
1. Complete the `get_workflow()` so that the `workflowresponse` also has serialized function attached to each operator. To get the serialized function, I route "/api/function/%s/export" and request by operator_id
2. Modify the previous `flow.get()` into `flow.artifact()`. In `flow.artifact()`, we first fetch the `artifact.Artifact` object with matching `artifact_name` and convert it to `generic_artifact.Artifact` so that we can call `.get()`.
3. For the integration test. I fetch the artifact using `flow.artifact(output_artifact.name())`. And check for name equality and dataframe equality between locally created artifact and fetched from server artifact.

File: sdk/aqueduct/api_client.py
Comment (kenxu95) [inline sdk/aqueduct/api_client.py:286]: instead of `change_file()` on the operator, I'd like to keep our current abstraction where the dag class is the only thing that can modify underlying operator and artifact objects. Having fewer places where things can modify the dag will make things easier to maintain + us to have fewer bugs.

Would you be able to use `dag.update_operator_spec()` instead? You can probably write a helper to construct the new operator spec from the old one + serialized contents.

File: sdk/aqueduct/api_client.py
Comment (kenxu95) [inline sdk/aqueduct/api_client.py:281]: typo: `work_flow`

Also, for better abstraction, the API client should work at a lower layer. It shouldn't know anything about our dag operators, it should only understand backend response objects. I would recommend exposing a separate `export` function in the APIClient and moving the dag updating logic up into the caller.

File: sdk/aqueduct/flow.py
Comment (kenxu95) [inline sdk/aqueduct/flow.py:182]: nit: lets make all these `elif`?

File: sdk/aqueduct/flow.py
Comment (kenxu95) [inline sdk/aqueduct/flow.py:166]: I think this `artifact()` method should be on the FlowRun object, not the flow. Always fetching from the latest run seems like an unnecessary restriction. A user should be able to fetch any historical run, and realize the artifact from that historical run, using latest data.

**Review by kenxu95 — CHANGES_REQUESTED** [review]
This is awesome! I have a number of comments but thanks for getting this working.

There are two other things I thought of while reviewing:
1) Did you want to rename the generic artifact to `GenericArtifact`? We can probably just do it (here or in a separate PR)
2) A returned artifact currently cannot be used to construct a new dag right? Because the underlying dag objects are different. Could we add a defensive check in there for now, where if you attempt to use a returned artifact in new computation, we'd throw an error? This might be an interesting follow up :)

File: sdk/aqueduct/api_client.py
Comment (Fanjia-Yan) [reply sdk/aqueduct/api_client.py:286]: @kenxu95 Since `get_workflow` returns a class `WorkflowDagResponse`, the serialize function is inserted into `WorkflowDagResponse` rather than the `dag` class. Therefore I believe the update_operator_spec() should be under  `WorkflowDagResponse`?

**Comment by Fanjia-Yan** [comment]
1. Migrate `flow.artifact()` to `flowrun.artifact()`
2. Create a `export_serialized_function()` to handle the getting serialized function into the dag_response rather than putting them all in `get_workflow()`
3. As mentioned by Kenny, the operation on `operator` is crossing the the abstraction barrier. Therefore, in the `api_client`, I eliminated any operation on `operator`  and put them into `workflow_dag_response`
4. In order to prevent reuse of `artifact` from `flowrun.arifact()`, I create a boolean parameter for Artifact(TableArtifact etc.) `from_flow_run`. The parameter defaults to false but can be set to true when creating the Artifact Object. Then, in `decorator.py`, when the operators and artifacts are wrapped, I check whether any input artifact `from_flow_run` is set to true. If so return a Exception.
5. I also create a test case to test whether we catch any input artifact from flow run. I basically publish a flow and fetch the artifact and reuse it for sentiment model. When feeding in the return artifact, we are expected an Exception.

File: sdk/aqueduct/api_client.py
Comment (kenxu95) [inline sdk/aqueduct/api_client.py:341]: ok this is better, but I was really saying this APIClient layer should be as thin as possible - it should literally just take in a single operator id and return the serialized function for that operator. The double-for loop belongs in the caller.

Question: does an operator always only correspond to a single workflow dag? Is there a case where an operator corresponds to multiple workflow dags on the backend. cc @cw75

File: sdk/aqueduct/check_artifact.py
Comment (kenxu95) [inline sdk/aqueduct/check_artifact.py:41]: Can you write a comment here describing what this variable represents? Its not super clear from the name in what scenarios it is `_from_flow_run` and what scenarios it isn't.

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:73]: nit: should be Optional[Union[...]], get rid of the None

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:72]: super nit: just "name" is fine. The function is called `artifact()` already.

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:82]: nit: uncapitalize DAG here, since we don't name like this anywhere else.

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:80]: this fetching of the artifact from the dag should be a method on the dag class (see dag.py). You can either augment `must_get_artifact()` or write a separate method called `get_artifact_by_name()` or something - see how all the other functions on this class do it. You'll probably have to define an `artifact_by_name` dictionary to do the lookup (similar to how `operator_by_name` works).

It should be simple here as:
```
artifact = self._dag.must_get_artifact(name)
```

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:92]: nit: better to use `get_artifact_type()` in these if-else statements, since its more readable and catches any non-valid types.

File: sdk/aqueduct/generic_artifact.py
Comment (kenxu95) [inline sdk/aqueduct/generic_artifact.py:12]: why is this necessary?

File: sdk/aqueduct/responses.py
Comment (kenxu95) [inline sdk/aqueduct/responses.py:158]: wait no, I thought we agreed on putting this write on the dag class?

File: sdk/aqueduct/operators.py
Comment (kenxu95) [inline sdk/aqueduct/operators.py:187]: I thought we agreed to use `update_operator_spec()` in dag.py? Not this.

File: sdk/aqueduct/decorator.py
Comment (kenxu95) [inline sdk/aqueduct/decorator.py:85]: This should be a `UserAction...Exception`. Look at the specific errors we have defined in `errors.py`

**Review by kenxu95 — CHANGES_REQUESTED** [review]
It's getting better! Thanks for fixing the incompatible dag issue. There are a lot of code quality and abstraction issues still here so I still need to block. I think maybe there was miscommunication about what abstractions are present in the SDK? Especially around how we update the dag.

**Comment by Fanjia-Yan** [comment]
After last review, several things have been changed:
1. the process of updating operator will be taking place in `dag.py` instead of `api_client.py`
2. the action of putting serialized function will be taking place in `flow.py: construct_flow_run()` instead of `api_client.py: get_workflow()`
3. fixing some misc and updating documentations for `_from_flow_run` parameter for InputArtifact
This version looks nicer since most codes are taken away from the api_client file :)

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:79]: I wonder if it might be nicer for the user if we return None if the artifact doesn't exist? Maybe they just want to check for existence of an artifact, and its a bit less annoying to check for None than it is to catch an error.

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:89]: Here I actually think we should raise an exception (aka fail very loudly), since this means something went wrong in our system. Is there an Internal error we can throw?

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [reply sdk/aqueduct/flow_run.py:79]: If you add that, could you also put that in the function comment string?

File: sdk/aqueduct/operators.py
Comment (kenxu95) [inline sdk/aqueduct/operators.py:187]: nit: no need to return None here. -> None is just a function without a return.

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [inline sdk/aqueduct/flow_run.py:87]: nit: for each of the artifact constructors, "True" is not a super enlightening argument value, so I typically do eg:
```
ParamArtifact(self._api_client, self._dag, artifact_from_dag.id, from_flow_run=True)
```
especially if its one of the last few arguments.

**Review by kenxu95 — APPROVED** [review]
Awesome! Thanks for pushing this through 👍 Left a few stylistic/cleanup comments but looks good to me. I think @vsreekanti also needs to unblock.

**Comment by vsreekanti** [comment]
Hey @Fanjia-Yan, Kenny knows this codebase much better than me, so if he's signed off on the PR, you should be good to go. 🙂 No need to wait for my approval here.

File: sdk/aqueduct/flow_run.py
Comment (Fanjia-Yan) [reply sdk/aqueduct/flow_run.py:89]: @kenxu95 I think this error is really unlikely trigger as an artifact falls in either the four categories. However, if there will be an error, I think it is likely to be either "InvalidArtifactError"(does not exist yet in error.py) or "UnprocessableEntityError"(Exception raised for errors occured when the Aqueduct system fails to process certain inputs.)

Comment (Fanjia-Yan) [reply sdk/aqueduct/flow_run.py:89]: Or maybe we can throw the general AqueductError?

File: sdk/aqueduct/flow_run.py
Comment (Fanjia-Yan) [reply sdk/aqueduct/flow_run.py:79]: I think that's a good idea!

File: sdk/aqueduct/flow_run.py
Comment (kenxu95) [reply sdk/aqueduct/flow_run.py:106]: Let's throw an `InternalAqueductError`. The reason that this needs to be internal is because its definitely something thats gone wrong in our system, and is not the user's fault.

This error should never happen, as you pointed out, but it's important in our code to make sure that we cover all assumptions. If any assumption breaks, we should also fail loudly so that the person that broke the assumption can fix the issue. In this case, the assumption is that the artifact is one of four types, but this can easily be changed (imagine someone adds a new type on the backend, but doesn't find this function and doesn't update it).

**Pull request merged**
